firedrake/mg/embedded.py

```diff
@@ -0,0 +1,300 @@
+import firedrake
+import ufl
+from firedrake.petsc import PETSc
+
+
+__all__ = ("EmbeddedDGTransfer", )
+
+
+native = frozenset(["Lagrange", "Discontinuous Lagrange", "Real", "Q", "DQ"])
+
+
+def EmbeddedDGTransfer(element, use_fortin_interpolation=True):
+    """Return an object that provides grid transfers.
+
+    This object provides transfers of functions between levels in a
+    mesh hierarchy for element types that are not supported directly
+    by Firedrake's multilevel implementation.  It computes transfers
+    by embedding the element in an appropriate DG space, transferring
+    the resulting DG function, and then mapping back.
+
+    :arg element: The UFL element that will be transferred.
+    :arg use_fortin_interpolation: If True, use Fortin operator
+       as an approximate mass inverse, otherwise, use an actual mass
+       solve.
+    """
+    if type(element) is firedrake.MixedElement:
+        return MixedTransfer(element, use_fortin_interpolation=use_fortin_interpolation)
+    else:
+        return SingleTransfer(element, use_fortin_interpolation=use_fortin_interpolation)
+
+
+class NativeTransfer(object):
+    """A transfer object that just calls the native transfer
+    routines."""
+    @staticmethod
+    def prolong(c, f):
+        return firedrake.prolong(c, f)
+
+    @staticmethod
+    def inject(f, c):
+        return firedrake.inject(f, c)
+
+    @staticmethod
+    def restrict(f, c):
+        return firedrake.restrict(f, c)
+
+
+class SingleTransfer(object):
+    """Create a transfer object for a single (not mixed) element."""
+    def __new__(cls, element, use_fortin_interpolation=True):
+        if element.family() in native:
+            return NativeTransfer
+        else:
+            return super().__new__(cls)
+
+    def __init__(self, element, use_fortin_interpolation=True):
+        degree = element.degree()
+        cell = element.cell()
+        shape = element.value_shape()
+        if len(shape) == 0:
+            DG = ufl.FiniteElement("DG", cell, degree)
+        elif len(shape) == 1:
+            shape, = shape
+            DG = ufl.VectorElement("DG", cell, degree, dim=shape)
+        else:
+            DG = ufl.TensorElement("DG", cell, degree, shape=shape)
+
+        self.embedding_element = DG
+        self.use_fortin_interpolation = use_fortin_interpolation
+        self._V_DG_mass = {}
+        self._DG_inv_mass = {}
+        self._V_approx_inv_mass = {}
+        self._V_inv_mass_ksp = {}
+        self._DG_work = {}
+        self._work_vec = {}
+        self._V_dof_weights = {}
+
+    def V_dof_weights(self, V):
+        """Dof weights for Fortin projection.
+
+        :arg V: function space to compute weights for.
+        :returns: A PETSc Vec.
+        """
+        key = V.dim()
+        try:
+            return self._V_dof_weights[key]
+        except KeyError:
+            f = firedrake.Function(V)
+            firedrake.par_loop(
+                """for (int i = 0; i < A.dofs; i++)
```

Comment:
Can you add a comment explaining what this `par_loop` is doing?

Reply:
This is just spinning over all the cells (this process owns) and adding one to every dof that the cell can see.  After halo exchange this has computed the dof weights.

Reply:
Yes, I meant in the code.

Reply:
Done in ecbbeb8b